### PR TITLE
Moving datagen options to `export`

### LINKED ISSUE #3409: Where should locale filtering take place in the data exporter?
Datagen takes the `--locales` CLI or API argument. Each exporter implements `IterableDataProvider` based on all locales for which it has data (such as in CLDR-JSON). These sets of locales may be fully, partially, or non-overlapping. We should discuss the right place to perform this filtering.

See https://github.com/unicode-org/icu4x/pull/3386#discussion_r1181181433

### PR DESCRIPTION
**Pull request by robertbastian**
This also necessarily moves the locale filtering from `supported_locales` into `export`, and resolves #3409.

Part of #3487.

Resolves most todos in https://github.com/unicode-org/icu4x/pull/3640#issuecomment-1662683203

### CODE REVIEW

**Comment by Manishearth** [comment]
(letting shane review locale filtering, I can but I'm less familiar with the code)

**Review by sffc — DISMISSED** [review]
Time for me to resolve conflicts on my side

**Review by sffc — APPROVED** [review]

File: provider/datagen/src/options.rs [hunk 1]
```diff
@@ -42,12 +42,11 @@ use std::collections::HashSet;
 #[non_exhaustive]
 #[derive(Debug, Clone, PartialEq, Default)]
 pub struct Options {
+    /// The set of keys to generate. See [`icu_datagen::keys`],
+    /// [`icu_datagen::all_keys`], [`icu_datagen::key`] and [`icu_datagen::keys_from_bin`].
+    pub keys: HashSet<icu_provider::DataKey>,
```
[the remark this answers is not in the document]
Comment (robertbastian) [reply provider/datagen/src/options.rs:52]: The `locales`, `segmentation_models`, and `collations` fields are collections already.

[the remark this answers is not in the document]
Comment (sffc) [reply provider/datagen/src/options.rs:52]: (Yeah, which is why my comment is "observation pre-existing")

We should probably discuss this choice especially before we ship this API

**Pull request merged**